docker/generate_makefile.py

```diff
@@ -15,189 +15,130 @@
+from common import fuzzer_utils
+from experiment.build import docker_images
 
-    # Generate rules for building/pulling all target/benchmark pairs.
-    all_build_targets = ' '.join(
-        ['build-{0}-{1}'.format(fuzzer, benchmark) for benchmark in benchmarks])
-    all_pull_targets = ' '.join(
-        ['pull-{0}-{1}'.format(fuzzer, benchmark) for benchmark in benchmarks])
-    print('build-{fuzzer}-all: {all_targets}'.format(
-        fuzzer=fuzzer, all_targets=all_build_targets))
-    print('pull-{fuzzer}-all: {all_targets}'.format(
-        fuzzer=fuzzer, all_targets=all_pull_targets))
+BASE_TAG = "gcr.io/fuzzbench"
+BENCHMARK_DIR = benchmark_utils.BENCHMARKS_DIR
+
+
+def _print_benchmark_fuzz_target(benchmarks):
+    """Prints benchmark variables from benchmark.yaml files."""
+    for benchmark in benchmarks:
+        benchmark_vars = yaml_utils.read(
+            os.path.join(BENCHMARK_DIR, benchmark, 'benchmark.yaml'))
+        print(benchmark + '-fuzz-target=' + benchmark_vars['fuzz_target'])
+        print()
+
+
+def _print_makefile_run_template(image):
+    fuzzer, benchmark = image['tag'].split('/')[1:]
+
+    for run_type in ('run', 'debug', 'test-run'):
+        print(('{run_type}-{fuzzer}-{benchmark}: ' +
+               '.{fuzzer}-{benchmark}-runner').format(run_type=run_type,
+                                                      benchmark=benchmark,
+                                                      fuzzer=fuzzer))
+
+        print('\
+\tdocker run \\\n\
+\t--cpus=1 \\\n\
+\t--cap-add SYS_NICE \\\n\
+\t--cap-add SYS_PTRACE \\\n\
+\t-e FUZZ_OUTSIDE_EXPERIMENT=1 \\\n\
+\t-e FORCE_LOCAL=1 \\\n\
+\t-e TRIAL_ID=1 \\\n\
+\t-e FUZZER={fuzzer} \\\n\
+\t-e BENCHMARK={benchmark} \\\n\
+\t-e FUZZ_TARGET=$({benchmark}-fuzz-target) \\\
+'.format(fuzzer=fuzzer, benchmark=benchmark))
+
+        if run_type == 'test-run':
+            print('\t-e MAX_TOTAL_TIME=20 \\\n\t-e SNAPSHOT_PERIOD=10 \\')
+        if run_type == 'debug':
+            print('\t-entrypoint "/bin/bash" \\\n\t-it ', end='')
+        else:
+            print('\t', end='')
+
+        print(os.path.join(BASE_TAG, image['tag']))
+        print()
+
+
+# TODO(tanq16): Add unit test.
+def _print_rules_for_image(name, image):
+    """Print makefile section for given image to stdout."""
+    if not ('base' in name or 'dispatcher' in name):
+        print('.', end='')
+    print(name + ':', end='')
+    if 'depends_on' in image:
+        for dep in image['depends_on']:
+            if 'base' in dep:
+                print(' ' + dep, end='')
+            else:
+                print(' .' + dep, end='')
+    print()
+    print('\tdocker build \\')
+    print('\t--tag ' + os.path.join(BASE_TAG, image['tag']) + ' \\')
+    print('\t--build-arg BUILDKIT_INLINE_CACHE=1 \\')
+    print('\t--cache-from ' + os.path.join(BASE_TAG, image['tag']) + ' \\')
+    if name == 'base-builder':
+        print('\t--cache-from ' + 'gcr.io/oss-fuzz-base/base-clang \\')
```

Comment:
why not cache from gcr.io/fuzzbench/base-builder ? that should include both gcr.io/oss-fuzz-base/base-clang and the stuff on top of it, no?

Reply:
I added as per [build.mk](https://github.com/google/fuzzbench/blob/master/docker/build.mk). I'll remove this.